instance/models/mixins/secret_keys.py

```diff
@@ -0,0 +1,105 @@
+# -*- coding: utf-8 -*-
+#
+# OpenCraft -- tools to aid developing and hosting free software projects
+# Copyright (C) 2015-2016 OpenCraft <contact@opencraft.com>
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+"""
+Instance app model mixins - Database
+"""
+
+# Imports #####################################################################
+
+import os
+import hashlib
+import hmac
+from base64 import b64encode, b64decode
+
+from django.db import models
+from django.template import loader
+
+# Functions ###################################################################
+
+
+def generate_secret_key(char_length):
+    """
+    Creates a base64-encoded string of character length `char_length`
+    for use in other functions.
+    """
+    if char_length <= 0 or char_length % 4:
+        # b64encode will always pad its output to four characters, regardless
+        # of the byte count. We'd rather error out than encode too few or too many.
+        raise ValueError('char_length must be a positive multiple of 4')
+    # We need three random bytes per four random b64 characters.
+    random_bytes = os.urandom(int(char_length * 0.75))
+    # Encode to a b64 string...
+    random_string = b64encode(random_bytes)
+    # ...and return!
+    return random_string
+
+
+# Classes #####################################################################
+
+class SecretKeyInstanceMixin(models.Model):
+    """
+    An instance that needs to generate a set of secret keys for different purposes
+    """
+    secret_key_b64encoded = models.CharField(max_length=48, blank=True)
+
+    class Meta:
+        abstract = True
+
+    def set_random_key(self):
+        """
+        Generates a random seed for this instance to use when creating
+        derived secret keys for instance variables
+        """
+        self.secret_key_b64encoded = generate_secret_key(48)
+
+    def get_secret_key_settings(self):
+        """
+        Render the secret keys template as YAML and return it for use
+        in the appserver.
+        """
+        if not self.secret_key_b64encoded:
+            return ''
```

Comment:
>Or were you thinking it's better not to break existing instances, which is likely what changing their secret key would do?

This is exactly what my thinking was. We have tests to ensure that new instances get a secure base key, but we don't want to break existing instances by either failing to create new appservers or by messing with the existing (missing) keys.